daprdocs/content/en/developing-applications/mcp/mcp-authentication.md

```diff
@@ -0,0 +1,197 @@
+
+While MCP servers lack identity, OAuth2 is a well established standard that can be used to properly authenticate MCP clients to MCP servers.
+
+OAuth2 becomes essential when MCP servers are:
+
+* Multi-tenant
+* Remote
+* Cloud-hosted
+* Connected to confidential systems
+* Performing privileged actions on behalf of a user
+* Exposing tools that must be permission-gated
+
+Dapr enables seamless OAuth2 authentication between MCP clients and servers using [middleware]({{% ref "middleware" %}}) components.
+
+## Types of authentication
+
+Dapr supports two critical authentication mechanisms for production grade deployments of MCP servers - Client-side and Server-side.
+
+### Client-side Authentication
+
+The client initiates OAuth2 to obtain an access token and includes it when connecting to the MCP server.
+This proves the user’s identity and permissions and is required for remote, sensitive, or multi-tenant MCP servers.
+It ensures the server can trust who is calling and what scopes the client is allowed to use.
+
+### Server-side Authentication
+
+The server validates the client’s token or, if missing or insufficient, triggers an OAuth2 login or scope upgrade.
+This is needed for cloud-hosted or shared MCP servers, tenant-aware systems, and integrations that require user-specific authorization.
+It enforces access control, isolates users, and protects privileged tools and data.
+
+## How to enable Client-side Authentication
+
+### Define the MCP Server as an HTTPEndpoint
+
+Dapr allows developers and operators to model remote HTTP services as resources that can be governed and invoked using the Dapr [Service Invocation API]({{% ref "service-invocation-overview" %}}).
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: HTTPEndpoint
+metadata:
+  name: "mcp-server"
+spec:
+  baseUrl: https://my-mcp-server:443
+  headers:
+  - name: "Accept"
+    value: "text/event-stream"
+```
+
+### Define the OAuth2 middleware and configuration components
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: oauth2
+spec:
+  type: middleware.http.oauth2
+  version: v1
+  metadata:
+  - name: clientId
+    value: "<client-id>"
+  - name: clientSecret
+    value: "<client-secret>"
+  - name: authURL
+    value: "<authorization-url>"
+  - name: tokenURL
+    value: "<token-url>"
+  - name: scopes
+    value: "<comma-separated scopes>"
+```
+
+Next, create the configuration component.
+
+```yaml
+piVersion: dapr.io/v1alpha1
+kind: Configuration
+metadata:
+  name: auth
+spec:
+  tracing:
+    samplingRate: "1"
+  httpPipeline:
+    handlers:
+    - name: oauth2 # reference the oauth component here
+      type: middleware.http.oauth2    
+```
+
+*Note: visit [this link]({{% ref "component-secrets.md" %}}) to read on how to provide secrets to Dapr components*
+
+### Call the MCP server using an MCP client
```

Comment:
Like the others, say create a file called "mcpclient.py" and copy the code below